### ISSUE TITLE
.NETCoreApp 2.0, System.IO.Compression assembly load failure

### ISSUE THREAD

**Issue by karolz-ms** [dotnet/corefx#18338]
I am doing app building using .NET Core 2.0.0-preview1-001961-00. At run time I am getting the following error:

System.IO.FileLoadException occurred
  HResult=0x80131040
  Message=Could not load file or assembly 'System.IO.Compression, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
  Source=System.Net.Http
  StackTrace:
   at System.Net.Http.WinHttpResponseParser.CreateResponseMessage(WinHttpRequestState state, Boolean doManualDecompressionCheck)
   at System.Net.Http.WinHttpHandler.<StartRequest>d__105.MoveNext()

As far as I can tell System.IO.Compression assembly is right next to System.Net.Http assembly that my application is loading, they are both in C:\Program Files\dotnet\shared\Microsoft.NETCore.App\2.0.0-preview1-001961-00

Not sure why there would be a mismatch?

**Comment by mellinoe** [comment]
> As far as I can tell System.IO.Compression assembly is right next to System.Net.Http assembly that my application is loading

What version is that assembly?

**Comment by karolz-ms** [comment]
System.IO.Compression, Version=4.2.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089

[assembly: AssemblyFileVersion("4.6.25210.01")]
[assembly: AssemblyInformationalVersion("4.6.25210.01 built by: dlab-DDVSOWINAGE042. Commit Hash: 56986cdb177869420d0ba0b6bc028134eceb333f")]

**Comment by karolz-ms** [comment]
Turns out this can be reproduced super-easily, just create a .NET 2.0 console application and run the following code

```csharp
using System;
using System.Net.Http;

namespace HttpTest
{
    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                HttpClient c = new HttpClient();
                Console.WriteLine("Starting request...");
                var response = c.GetAsync("http://www.microsoft.com/").GetAwaiter().GetResult();
                Console.WriteLine($"The result code is {response.StatusCode}");
            }
            catch(Exception e)
            {
                Console.WriteLine($"Error: {e.ToString()}");
            }

            Console.WriteLine("Press any key to exit");
            Console.ReadKey(intercept: true);
        }
    }
}
```

Project:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Net.Http" Version="4.3.1" />
  </ItemGroup>

</Project>
```

**Comment by karolz-ms** [comment]
Actually the error is slightly different. This time I am getting an assembly load error for System.Net.Http, version 4.2.0.

If I change the System.Net.Http dependency to 4.2.0, I get a Nuget warning during restore:
Restoring NuGet packages...
Dependency specified was System.Net.Http (>= 4.2.0) but ended up with System.Net.Http 4.3.0.

**Comment by karolz-ms** [comment]
Also tried 

```xml
<PackageReference Include="System.Net.Http" Version="[4.2.0, 4.3.0)" />
```

That results in NuGet restore failure

C:\Program Files\dotnet\sdk\2.0.0-preview1-005783\NuGet.targets(97,5): error : Unable to resolve 'System.Net.Http (>= 4.2.0 && < 4.3.0)' for '.NETCoreApp,Version=v2.0'. [c:\code\scratch\HttpTest\HttpTest.csproj]

**Comment by mellinoe** [comment]
I don't think you should have that package reference. Try removing it and things should work.

**Comment by karolz-ms** [comment]
OK, thanks. So much for the simple repro though... Tried to be helpful and just muddied the waters  :-(

I do not have a simple repro but I am still suffering from the issue with the app that I am developing. How can I diagnose this? Is there an equivalent of fuslogvw in the Core world?

**Comment by mellinoe** [comment]
If you are using a library that relies on that package, you may just need to add a newer version of the package (not 4.3.0). .NET Core 2.0 is in a prerelease state; you will probably need to reference a prerelease version of System.Net.Http as well, in order to get the correct asset for .NET Core 2.0.

cc: @ericstj

**Comment by karolz-ms** [comment]
OK, I was able to figure it out. One of my dependencies depends on System.Diagnostics.StackTrace 4.3.0, which depends on System.Reflection.Metadata 1.4.1, which depends on System.IO.Compression 4.3.0, which causes the problem.

I guess I will now need to find the owners of all these packages and ask for an update?

**Comment by ericstj** [comment]
You don't need a package reference to use either of these in NETCoreApp2.0.  They are part of the framework.  All you do by refrencing those older packages is excercise our conflict-resolution feature that will remove them from the app.

If there is any version mismatch it could indicate a problem with conflict resolution.  @dsplaisted just [fixed](https://github.com/dotnet/sdk/commit/aa0f30e4c4ee695751f01cda77b1c50bfbdb18ff) a build issue around this.  If you'd like to see if that fix would resolve your issue, try publishing and running the published output.  If it works on publish you probably hit that bug.

If you think it's not that bug, share a /flp:v=diag build log and I'll take a look.  FWIW I tried your simple repo and it worked for me.

**Comment by ericstj** [comment]
> One of my dependencies depends on System.Diagnostics.StackTrace 4.3.0, which depends on System.Reflection.Metadata 1.4.1, which depends on System.IO.Compression 4.3.0, which causes the problem.

That shouldn't be a problem.  Please see my last response.  I suspect this is related to conflict-resolution.

**Comment by karolz-ms** [comment]
Yes, I think that is it-- the published output seems to work. For reference, here is a better repro, reflecting the problem with the original app. It consists of two projects: a library and an "exe"

Library
```csharp
using System;
using System.Net.Http;

namespace HttpTestLib
{
    public class Lib
    {
        public static void DoRequest()
        {
            try
            {
                HttpClient c = new HttpClient();
                c.DefaultRequestHeaders.AcceptEncoding.Add(new System.Net.Http.Headers.StringWithQualityHeaderValue("gzip"));
                Console.WriteLine("Starting request...");
                var response = c.GetAsync("http://www.microsoft.com/").GetAwaiter().GetResult();
                Console.WriteLine($"The result code is {response.StatusCode}");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Error: {e.ToString()}");
            }
        }
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard1.3</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Diagnostics.StackTrace" Version="4.3.0" />
  </ItemGroup>

</Project>
```

Exe

```csharp
namespace HttpTest
{
    class Program
    {
        static void Main(string[] args)
        {
            HttpTestLib.Lib.DoRequest();

            Console.WriteLine("Press any key to exit");
            Console.ReadKey(intercept: true);
        }
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <ProjectReference Include="..\HttpTestLib\HttpTestLib.csproj" />
  </ItemGroup>

</Project>
```

When can I expect Daniel's fix to be available in .NET builds?

**Comment by ViktorHofer** [comment]
Fixed in Core 2.0 stable release.
